File: src/main/java/ca/mcgill/science/tepid/client/lpd/LPDServer.java
package ca.mcgill.science.tepid.client.lpd;

import ca.mcgill.science.tepid.models.data.PrintJob;

import java.io.*;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

public class LPDServer implements Closeable {
    private boolean closed;
    private final ServerSocket serverSock;
    private final Queue<LPDJobListener> listeners = new ConcurrentLinkedQueue<>();

    public LPDServer(int port) throws IOException {
        System.out.println("Creating LPDServer in port " + port);
        serverSock = new ServerSocket(port, 0, InetAddress.getByName(null));
    }

    public void start() {
        while (!closed) {
            try {
                new LPDClientHandler(serverSock.accept(), LPDServer.this).start();
                System.out.println("New lpd");
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void close() throws IOException {
        this.closed = true;
        System.out.println("Closing LPDServer");
        serverSock.close();
    }

    public void addJobListener(LPDJobListener l) {
        listeners.add(l);
    }

    protected void newJob(final PrintJob p, InputStream is, long size) {
        if (size < 1) size = Long.MAX_VALUE;
        Queue<PipedOutputStream> streams = new ConcurrentLinkedQueue<>();
        for (final LPDJobListener l : listeners) {
            try {
                PipedOutputStream pos = new PipedOutputStream();
                final PipedInputStream pis = new PipedInputStream(pos);
                streams.add(pos);
                new Thread("Print Job Event") {
                    @Override
                    public void run() {
                        l.printJob(p, pis);
                        try {
                            pis.close();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                }.start();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        try {
            int bytesRead, totalBr = 0;
            byte[] buf = new byte[2048];
            while (totalBr < size && (bytesRead = is.read(buf)) > 0) {
                for (OutputStream os : streams) {
                    os.write(buf, 0, bytesRead);
                }
                totalBr += bytesRead;
            }
            for (OutputStream os : streams) os.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: src/main/java/ca/mcgill/science/tepid/client/lpd/LPDServer.kt
package ca.mcgill.science.tepid.client.lpd

import ca.mcgill.science.tepid.models.data.PrintJob
import ca.mcgill.science.tepid.utils.WithLogging
import java.io.*
import java.net.InetAddress
import java.net.ServerSocket
import java.util.concurrent.*

class LPDServer(port: Int) : WithLogging(), Closeable {
    private var closed: Boolean = false

    init {
        log.info("Creating LPDServer in port $port")
    }

    private val serverSock: ServerSocket = ServerSocket(port, 0, InetAddress.getByName(null))
    private val listeners = ConcurrentLinkedQueue<(PrintJob, InputStream) -> Unit>()
    private val executor: ExecutorService = ThreadPoolExecutor(1, 10, 1, TimeUnit.MINUTES,
            ArrayBlockingQueue<Runnable>(30, true))

    fun start() {
        while (!closed) {
            try {
                LPDClientHandler(serverSock.accept(), this@LPDServer).start()
                log.info("New lpd")
            } catch (e: IOException) {
                log.error(e)
            }
        }
    }

    @Synchronized
    override fun close() {
        this.closed = true
        log.info("Closing LPDServer")
        try {
            listeners.clear()
            serverSock.close()
        } catch (e: IOException) {
            log.error("Failed to dispose", e)
        }
    }

    fun addJobListener(listener: (job: PrintJob, input: InputStream) -> Unit) {
        listeners.add(listener)
    }

    fun newJob(p: PrintJob, input: InputStream, size: Int) {
        val streams = ConcurrentLinkedQueue<PipedOutputStream>()
        log.info("Starting job events")
        for (l in listeners) {
            try {
                val pos = PipedOutputStream()
                val pis = PipedInputStream(pos)
                streams.add(pos)
                executor.execute {
                    pis.use { l(p, it) }
                }
            } catch (e: IOException) {
                log.error("Failed to run job listener", e)
            }
        }
        log.info("Bound job listeners")
        var bytesRead: Int
        var totalBr = 0
        val maxSize = if (size >= 1) size else Int.MAX_VALUE
        val buf = ByteArray(2048)
        try {
            while (totalBr < maxSize) {
                bytesRead = input.read(buf)
                if (bytesRead <= 0) break
                streams.forEach { it.write(buf, 0, bytesRead) }
                totalBr += bytesRead
            }
        } catch (e: IOException) {
            log.info("Error in passing pipe streams", e)
        } finally {
            streams.forEach(PipedOutputStream::close)
        }
    }
}
